File: VBNetCore.Security/UserPrincipal.vb
Imports System.Linq
Imports System.Security.Principal

Namespace VBNetCore.Security
    Public Class UserPrincipal
        Implements IPrincipal
        Private _identity As UserIdentity
        Public Property Identity As UserIdentity
            Get
                Return If(_identity, Function()
                                         _identity = New AnonymousIdentity()
                                         Return _identity
                                     End Function())
            End Get
            Set(ByVal value As UserIdentity)
                _identity = value
            End Set
        End Property
        Private ReadOnly Property PIdentity As IIdentity Implements IPrincipal.Identity
            Get
                Return Identity
            End Get
        End Property

        Public Function IsInRole(ByVal role As String) As Boolean Implements IPrincipal.IsInRole
            Return _identity.Roles.Contains(role)
        End Function
        Public Function IsInPermission(ByVal permission As String) As Boolean
            Return _identity.Permissions.Contains(permission)
        End Function
    End Class
End Namespace

File: VBNetCore.Security/UserPrincipal.cs
using System.Linq;
using System.Security.Principal;

namespace NetCore.Security
{
    public class UserPrincipal : IPrincipal
    {
        private UserIdentity _identity;
        public UserIdentity Identity
        {
            get => _identity ?? (_identity = new AnonymousIdentity());
            set => _identity = value;
        }
        IIdentity? IPrincipal.Identity => Identity;

        public bool IsInRole(string role)
        {
            return _identity.Roles.Contains(role);
        }
        public bool IsInPermission(string permission)
        {
            return _identity.Permissions.Contains(permission);
        }
    }
}
